providers/apis/defi/ethmulticlient/multi_client.go

```diff
@@ -87,77 +89,103 @@ func NewMultiRPCClientFromEndpoints(
+				)
+			} else { // the batch call succeeded, and the eth_blockNumber call had results.\
+				// try to get the block number.
+				r, ok := req[blockNumReqIndex].Result.(*string)
+				if !ok {
+					errs[i] = fmt.Errorf("result from eth_blockNumber was not a string")
+					m.logger.Debug(
+						"result from eth_blockNumber was not a string",
+						zap.String("url", url),
+					)
+				} else {
+					// decode the new height
+					height, err := hexutil.DecodeUint64(*r)
+					if err != nil { // if we can't decode the height, log an error.
+						errs[i] = fmt.Errorf("could not decode hex eth height: %w", err)
+						m.logger.Debug(
+							"could not decode hex eth height",
+							zap.String("url", url),
+							zap.Error(err),
+						)
+					} else { // if we _can_ decode the height, append these results.
+						m.logger.Debug(
+							"got height for eth batch request",
+							zap.Uint64("height", height),
+							zap.String("url", url),
+						)
+						// append the results, minus the appended eth_blockNumber request.
+						results[i] = result{height, req[:blockNumReqIndex]}
+					}
+				}
+			}
+			wg.Done()
+		}(clientIdx)
+	}
+	wg.Wait()
 
-	// TODO(david): consider parallelizing these requests.
+	// see which of the results had the largest height, and store the index of that result.
 	var maxHeight uint64
-	for i, client := range m.clients {
-		url := m.api.Endpoints[i].URL
-
-		err := client.BatchCallContext(ctx, req)
-		if err != nil || req[blockNumReqIndex].Result == "" || req[blockNumReqIndex].Error != nil {
-			errs = fmt.Errorf("%w: endpoint request failed: %w, %w", errs, err, req[blockNumReqIndex].Error)
-			m.logger.Debug(
-				"endpoint request failed",
-				zap.Error(err),
-				zap.Any("result", req[blockNumReqIndex].Result),
-				zap.Error(req[blockNumReqIndex].Error),
-				zap.String("url", url),
-			)
-			continue
-		}
-
-		r, ok := req[blockNumReqIndex].Result.(*string)
-		if !ok {
-			errs = fmt.Errorf("%w: result from eth_blockNumber was not a string", errs)
-			m.logger.Debug(
-				"result from eth_blockNumber was not a string",
-				zap.String("url", url),
-			)
-			continue
-		}
-
-		newHeight, err := hexutil.DecodeUint64(*r)
-		if err != nil {
-			errs = fmt.Errorf("%w: could not decode hex eth height: %w", errs, err)
-			m.logger.Debug(
-				"could not decode hex eth height",
-				zap.String("url", url),
-				zap.Error(err),
-			)
-			continue
-		}
-		m.logger.Debug(
-			"got height for eth batch request",
-			zap.Uint64("height", newHeight),
-			zap.String("url", url),
-		)
-
-		if newHeight > maxHeight {
-			m.logger.Debug("new max eth height seen",
-				zap.Uint64("prev_height", maxHeight),
-				zap.Uint64("new_height", newHeight),
-				zap.String("url", url),
-			)
-
-			maxHeight = newHeight
-			copy(batchElems, req[:blockNumReqIndex])
+	var maxHeightIndex int
+	for i, res := range results {
+		if res.height > maxHeight {
+			maxHeight = res.height
+			maxHeightIndex = i
 		}
 	}
```

Comment:
i considered doing a buffered channel and analyzing the results on the fly, but seemed potentially overkill. gathering the results then doing a once over seemed more straightforward. happy to be wrong here though. just lmk your thoughts <3

addtly, anyone know the typical amt of clients a multiclient might have?